topic-operator/src/main/java/io/strimzi/operator/topic/Session.java

```diff
@@ -171,96 +216,116 @@ public void start(Promise<Void> start) {
         String clientId = config.get(Config.CLIENT_ID);
         LOGGER.debug("Using client-Id {}", clientId);
 
-        Zk.create(vertx, config.get(Config.ZOOKEEPER_CONNECT),
-                this.config.get(Config.ZOOKEEPER_SESSION_TIMEOUT_MS).intValue(),
-                this.config.get(Config.ZOOKEEPER_CONNECTION_TIMEOUT_MS).intValue(),
-            zkResult -> {
-                if (zkResult.failed()) {
-                    start.fail(zkResult.cause());
-                    return;
-                }
-                this.zk = zkResult.result();
-                LOGGER.debug("Using ZooKeeper {}", zk);
-
-                String topicsPath = config.get(Config.TOPICS_PATH);
-                TopicStore topicStore;
-                if (config.get(Config.USE_ZOOKEEPER_TOPIC_STORE)) {
-                    topicStore = new ZkTopicStore(zk, topicsPath);
-                } else {
-                    boolean exists = zk.getPathExists(topicsPath);
-                    CompletionStage<KafkaStreamsTopicStoreService> cs;
-                    if (exists) {
-                        cs = Zk2KafkaStreams.upgrade(zk, config, adminClientProperties(), false);
-                    } else {
-                        KafkaStreamsTopicStoreService ksc = new KafkaStreamsTopicStoreService();
-                        cs = ksc.start(config, adminClientProperties()).thenCompose(s -> CompletableFuture.completedFuture(ksc));
-                    }
-                    topicStore = ConcurrentUtil.result(
-                            cs.handle((s, t) -> {
-                                if (t != null) {
-                                    LOGGER.error("Failed to create topic store.", t);
-                                    start.fail(t);
-                                    return null; // [1]
-                                } else {
-                                    service = s;
-                                    return s.store;
-                                }
-                            }));
-                    if (topicStore == null) {
-                        return; // [1]
-                    }
-                }
+        startHealthServer(topicOperatorState)
+                .onFailure(cause -> LOGGER.error("Failed to start health server", cause))
+                .onSuccess(httpServer -> healthServer = httpServer)
+                .compose(ignored -> zkClientCreator.apply(vertx, config))
+                .onSuccess(zookeeper -> zk = zookeeper)
+                .compose(zk -> createTopicStoreAsync(zk, config))
+                .onSuccess(topicStore -> LOGGER.debug("Using TopicStore {}", topicStore))
+                .compose(topicStore -> createTopicOperatorAndZkWatchers(labels, namespace, topicStore))
+                .compose(this::createK8sWatcher)
```

Comment:
Looking at `startWatcher` (which is called by `createK8sWatcher`)... it looks like it could block in the kube client.

Reply:
Cheers @tombentley, I'll guard against that.

Reply:
I've added a unit test to verify that the changes work.:)